osquery/tables/system/windows/registry.h

```diff
@@ -0,0 +1,160 @@
+static LONGLONG filetimeToUnixtime(FILETIME ft) {
+  LARGE_INTEGER date, adjust;
+  date.HighPart = ft.dwHighDateTime;
+  date.LowPart = ft.dwLowDateTime;
+  adjust.QuadPart = 11644473600000 * 10000;
+  date.QuadPart -= adjust.QuadPart;
+  return date.QuadPart / 10000000;
+};
+
+/// Microsoft helper function for getting the contents of a registry key
+std::vector<std::map<std::string, std::string>> queryKey(HKEY hKey) {
+  TCHAR achClass[MAX_PATH] = TEXT("");
+  DWORD cchClassName = MAX_PATH;
+  DWORD cSubKeys = 0;
+  DWORD cbMaxSubKey;
+  DWORD cchMaxClass;
+  DWORD cValues;
+  DWORD cchMaxValueName;
+  DWORD cbMaxValueData;
+  DWORD cbSecurityDescriptor;
+  DWORD retCode;
+  FILETIME ftLastWriteTime;
+  retCode = RegQueryInfoKey(hKey,
+                            achClass,
+                            &cchClassName,
+                            nullptr,
+                            &cSubKeys,
+                            &cbMaxSubKey,
+                            &cchMaxClass,
+                            &cValues,
+                            &cchMaxValueName,
+                            &cbMaxValueData,
+                            &cbSecurityDescriptor,
+                            &ftLastWriteTime);
+
+  std::vector<std::map<std::string, std::string>> queryResults;
+  TCHAR achKey[MAX_KEY_LENGTH];
+  DWORD cbName;
+  if (cSubKeys > 0) {
+    for (auto i = 0; i < cSubKeys; i++) {
+      cbName = MAX_KEY_LENGTH;
+      retCode = RegEnumKeyEx(hKey, i, achKey, &cbName, nullptr, nullptr,
+                             nullptr, &ftLastWriteTime);
+      if (retCode != ERROR_SUCCESS) {
+        continue;
+      }
+      Row r;
+      r["reg_type"] = "subkey";
+      r["index"] = std::to_string(i);
+      r["subkey"] = achKey;
+      r["mtime"] = std::to_string(filetimeToUnixtime(ftLastWriteTime));
+      queryResults.push_back(r);
+    }
+  }
+
+  if (cValues <= 0) {
+    return queryResults;
+  }
+
+  BYTE* bpDataBuff = new BYTE[cbMaxValueData];
+  DWORD cchValue = MAX_VALUE_NAME;
+  TCHAR achValue[MAX_VALUE_NAME];
+
+  for (size_t i = 0, retCode = ERROR_SUCCESS; i < cValues; i++) {
+    std::vector<std::string> multiSzStrs;
+    size_t cnt = 0;
+    ZeroMemory(bpDataBuff, cbMaxValueData);
+    cchValue = MAX_VALUE_NAME;
+    achValue[0] = '\0';
+
+    retCode = RegEnumValue(hKey, i, achValue, &cchValue, nullptr, nullptr,
+                           nullptr, nullptr);
+
+    if (retCode != ERROR_SUCCESS) {
+      continue;
+    }
+
+    DWORD lpData = cbMaxValueData;
+    DWORD lpType;
+    LONG dwRes =
+        RegQueryValueEx(hKey, achValue, 0, &lpType, bpDataBuff, &lpData);
+    Row r;
+    r["reg_type"] = "value";
+    r["index"] = std::to_string(i);
+    r["name"] = achValue;
+    r["type"] = std::to_string(lpType);
+    r["mtime"] = std::to_string(filetimeToUnixtime(ftLastWriteTime));
+
+    switch (lpType) {
+    case REG_BINARY:
+      r["data"] = std::string(reinterpret_cast<char const*>(bpDataBuff));
```

Comment:
These `reinterpret_cast`s are scary, let's make sure there's no other way to get a `BYTE*` into a `const char*`. Maybe a [`dynamic_cast`](http://en.cppreference.com/w/cpp/language/dynamic_cast), which does provide error handling and is considered safe, yet slow.